bin/uniref90_non_viral_filter.py

```diff
@@ -0,0 +1,89 @@
+#!/usr/bin/env python
+# -*- coding: utf-8 -*-
+
+# Copyright 2024 EMBL - European Bioinformatics Institute
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+# http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import argparse
+import gzip
+
+from Bio import SeqIO, Entrez
+import taxoniq
+
+
+def is_virus_taxoniq(tax_id):
+    """
+    Check if taxa id is virus using faster taxoniq search in local, 
+    indexed, compressed copy of the NCBI taxonomy database
+    """
+    taxon = taxoniq.Taxon(tax_id)
+    if taxon.ranked_lineage:
+        highest_taxon = taxon.ranked_lineage[-1]
+        if highest_taxon.rank.name == "superkingdom":
+            return highest_taxon.scientific_name == "Viruses"
+    raise ValueError("No information about the lineage in taxoniq")
+
+
+def is_virus_entrez(taxid):
+    """
+    Check if taxa id is virus using slower approach with
+    Entrez quering
+    """
+    handle = Entrez.efetch(db="taxonomy", id=taxid, retmode="xml")
+    records = Entrez.read(handle)
+    handle.close()
+
+    lineage = records[0]["Lineage"]
+    highest_taxon = lineage.split("; ")[0]
+    return highest_taxon == "Viruses"
+
+
+def filter_fasta(in_handle, out_handle, err_handle):
+    """
+    Filter viral proteins from FASTA using TaxID field.
+    """
+    for record in SeqIO.parse(in_handle, "fasta"):
+        tax_id = record.description.split("TaxID=")[1].split()[0]
+        try:
+            try:
+                is_viral_protein = is_virus_taxoniq(tax_id)
+            except (KeyError, ValueError):
+                is_viral_protein = is_virus_entrez(tax_id)
+            if is_viral_protein is False:
+                SeqIO.write(record, out_handle, "fasta")
+        except Exception as e:
+            err_handle.write(f"Error while processing {tax_id}: {e}\n")
+
+
+def processing_handle(input_fasta, output_fasta):
+    """
+    Enable processing of both regular and gzipped FASTA files.
+    """
+    with open(output_fasta, 'w') as out_handle, open("failed.txt", "w") as err_handle:
+        if input_fasta.endswith('.gz'):
```

Comment:
I think biopython SeqIO automatically open gzip and non-gzip files, I think it's `SeqIO.parse`

Reply:
I can't find any examples when it works. 
If I try i fails with the error 
`UnicodeDecodeError: 'utf-8' codec can't decode byte 0x8b in position 1: invalid start byte`

Reply:
Yup... I was wrong, it doesn't support that (I was sure it could :face_exhaling:).
Reference docs -> https://biopython.org/docs/latest/Tutorial/chapter_seqio.html#sec-seqio-compressed